ledger/participant-integration-api/src/main/scala/platform/store/backend/common/CompletionStorageBackendTemplate.scala

```diff
@@ -6,42 +6,46 @@ package com.daml.platform.store.backend.common
 import java.sql.Connection
 import java.time.Instant
 
+import anorm.SqlParser.{binaryStream, int, str}
 import anorm.{RowParser, ~}
-import anorm.SqlParser.{int, str}
 import com.daml.ledger.api.v1.command_completion_service.CompletionStreamResponse
-import com.daml.ledger.api.v1.completion.Completion
 import com.daml.ledger.offset.Offset
 import com.daml.lf.data.Ref
 import com.daml.lf.data.Ref.Party
-import com.daml.platform.store.CompletionFromTransaction.toApiCheckpoint
+import com.daml.platform.store.CompletionFromTransaction
 import com.daml.platform.store.Conversions.{instant, offset}
 import com.daml.platform.store.backend.CompletionStorageBackend
-import com.google.rpc.status.Status
+import com.google.protobuf.CodedInputStream
+import com.google.rpc.status.{Status => StatusProto}
 
 trait CompletionStorageBackendTemplate extends CompletionStorageBackend {
 
   def queryStrategy: QueryStrategy
 
-  private val sharedCompletionColumns: RowParser[Offset ~ Instant ~ String] =
+  private val sharedColumns: RowParser[Offset ~ Instant ~ String] =
     offset("completion_offset") ~ instant("record_time") ~ str("command_id")
 
   private val acceptedCommandParser: RowParser[CompletionStreamResponse] =
-    sharedCompletionColumns ~ str("transaction_id") map {
+    sharedColumns ~ str("transaction_id") map {
       case offset ~ recordTime ~ commandId ~ transactionId =>
-        CompletionStreamResponse(
-          checkpoint = toApiCheckpoint(recordTime, offset),
-          completions = Seq(Completion(commandId, Some(Status()), transactionId)),
-        )
+        CompletionFromTransaction.acceptedCompletion(recordTime, offset, commandId, transactionId)
     }
 
-  private val rejectedCommandParser: RowParser[CompletionStreamResponse] =
-    sharedCompletionColumns ~ int("status_code") ~ str("status_message") map {
-      case offset ~ recordTime ~ commandId ~ statusCode ~ statusMessage =>
-        CompletionStreamResponse(
-          checkpoint = toApiCheckpoint(recordTime, offset),
-          completions = Seq(Completion(commandId, Some(Status(statusCode, statusMessage)))),
-        )
-    }
+  private val rejectedCommandParser: RowParser[CompletionStreamResponse] = {
+    val parserWithCodeAndMessage = sharedColumns ~
+      int("rejection_status_code") ~ str("rejection_status_message") map {
+        case offset ~ recordTime ~ commandId ~ rejectionStatusCode ~ rejectionStatusMessage =>
+          val status = StatusProto.of(rejectionStatusCode, rejectionStatusMessage, Seq.empty)
+          CompletionFromTransaction.rejectedCompletion(recordTime, offset, commandId, status)
+      }
```

Comment:
do you think we have tests somewhere to test this part?
normally it would not be tested because on the HEAD we would always populate the rejection_status

Reply:
@rautenrieth-da did you maybe encountered a test suit which would test this?

Reply:
We are very bad at testing migrations. I have no idea how we would go about testing this.

Reply:
The [compatibility package](https://github.com/digital-asset/daml/tree/main/compatibility) tests migrations in a way. It does:

1. install an old SDK version
2. push some data into a sandbox ledger, using the ledger API
3. upgrade to the next SDK version (up to and including the head version)
4. use the ledger API to check that the ledger content still looks the same
5. go to step 2

The application that uploads and checks data is a custom client application written in Haskell. There are checks for comparing the ACS and contract visibility due to divulgence. It could probably be extended to compare completions as well.

Reply:
Yeah but do we also test rejections this way? Never mind me, it will probably work...although we would not necessarily would have here dead code if we follow @miklos-da useful suggestion

Reply:
I don't think we test rejections, but we could extend the compatibility tests to cover them.